### PR TITLE
[SPARK-11602] [MLlib] Refine visibility for 1.6 scala API audit

### PR DESCRIPTION
**Pull request by hhbyyh**
jira: https://issues.apache.org/jira/browse/SPARK-11602 

Made a pass on the API change of 1.6. Open the PR for efficient discussion.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/SQLTransformer.scala [hunk 1]
```diff
@@ -44,7 +44,7 @@ class SQLTransformer @Since("1.6.0") (override val uid: String) extends Transfor
    * @group param
    */
   @Since("1.6.0")
-  final val statement: Param[String] = new Param[String](this, "statement", "SQL statement")
+  private final val statement: Param[String] = new Param[String](this, "statement", "SQL statement")
```
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/SQLTransformer.scala]: Actually a lot of Params in ml.feature is not private, even with setter and getter.

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/SQLTransformer.scala]: They should be public.  Users should be able to reference them in order to do things like create a ParamMap, or print the built-in Param doc.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeans.scala [hunk 2]
```diff
@@ -407,7 +407,7 @@ private object BisectingKMeans extends Serializable {
  */
 @Since("1.6.0")
 @Experimental
-class ClusteringTreeNode private[clustering] (
+private[clustering] class ClusteringTreeNode private[clustering] (
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeans.scala]: I think this was intended to be public.  @yu-iskw can you please confirm?

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala [hunk 3]
```diff
@@ -33,7 +33,7 @@ import org.apache.spark.rdd.RDD
 @Since("1.6.0")
 @Experimental
 class BisectingKMeansModel @Since("1.6.0") (
-    @Since("1.6.0") val root: ClusteringTreeNode
+    @Since("1.6.0")private[clustering] val root: ClusteringTreeNode
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: @yu-iskw Can you confirm if you meant this to be public?  It may not be necessary since most users will just call predict, computeCost, etc.  What do you think?  It does expose a much larger API.

File: mllib/src/main/scala/org/apache/spark/ml/feature/PCA.scala [hunk 4]
```diff
@@ -105,7 +105,7 @@ object PCA extends DefaultParamsReadable[PCA] {
 @Experimental
 class PCAModel private[ml] (
     override val uid: String,
-    val pc: DenseMatrix)
+    private[feature] val pc: DenseMatrix)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/PCA.scala]: This seems like a pretty stable API to expose.  I'd be OK with keeping it public.  The only reason not to would be if we might change pc to be a distributed matrix someday.  But I don't think we will since it's supposed to be fairly small (low-rank).

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala (hunk 3 above)
Comment (yu-iskw) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: Sorry. It's not public parameter.

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: No problem; thanks for confirming.

File: mllib/src/main/scala/org/apache/spark/ml/feature/PCA.scala (hunk 4 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/PCA.scala]: OK, Thanks. I'll revert this.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeans.scala [hunk 5]
```diff
@@ -407,7 +407,7 @@ private object BisectingKMeans extends Serializable {
...
     val index: Int,
     val size: Long,
     private val centerWithNorm: VectorWithNorm,
```
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeans.scala]: If ClusteringTreeNode should be public, how about the fields index, size, cost and children?

**Comment by hhbyyh** [comment]
This should be all for visibility review.

**Comment by jkbradley** [comment]
My apologies!  I lost track of this PR, but will check again now.  Could you please resolve the merge conflicts?

File: mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala [hunk 6]
```diff
@@ -52,6 +52,10 @@ private[feature] trait StringIndexerBase extends Params with HasInputCol with Ha
     val outputFields = inputFields :+ attr.toStructField()
     StructType(outputFields)
   }
+
+  /** @group setParam */
+  def setHandleInvalid(value: String): this.type = set(handleInvalid, value)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala]: This may break Java compatibility (setters in traits don't seem to work).  I'd revert the change.

File: mllib/src/main/scala/org/apache/spark/ml/feature/package-info.java [hunk 7]
```diff
@@ -23,7 +23,7 @@
  * features into more suitable forms for model fitting.
  * Most feature transformers are implemented as {@link org.apache.spark.ml.Transformer}s, which
  * transforms one {@link org.apache.spark.sql.DataFrame} into another, e.g.,
- * {@link org.apache.spark.feature.HashingTF}.
+ * {@link org.apache.spark.mllib.feature.HashingTF}.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/package-info.java]: Good catch, but should be "ml" instead of "mllib"

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 8]
```diff
@@ -529,7 +529,7 @@ class LinearRegressionSummary private[regression] (
     val predictionCol: String,
     val labelCol: String,
     val model: LinearRegressionModel,
-    val diagInvAtWA: Array[Double]) extends Serializable {
+    private val diagInvAtWA: Array[Double]) extends Serializable {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: This could be useful to users, but it should be documented if public.  @yanboliang Did you and @mengxr intend this to be public?

**Comment by jkbradley** [comment]
Done with pass

File: mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala (hunk 6 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala]: reverted

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala (hunk 8 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: @jkbradley I vote for private at present. Users will care about ```coefficientStandardErrors```, ```tValues``` and ```pValues``` which are calculated by ```diagInvAtWA``. Meanwhile users can not get ```diagInvAtWA``` at R by public interface. @mengxr

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala (hunk 3 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: Just realized we no longer need the Since tag, but it's not a big deal

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala [hunk 9]
```diff
@@ -33,7 +33,7 @@ import org.apache.spark.rdd.RDD
 @Since("1.6.0")
 @Experimental
 class BisectingKMeansModel @Since("1.6.0") (
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: Just realized this constructor should be private since the argument is private.  As long as you're updating this, could you please eliminate the now-unnecessary Since annotations here?

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala (hunk 8 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: I think it should be private.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala (hunk 3 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: If we want to hide `root`, we should also hide class `ClusteringTreeNode` entirely because it is not used anywhere else.

Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: Then I'll keep ClusteringTreeNode private for now.

**Comment by jkbradley** [comment]
LGTM.  I'll merge this with master and branch-1.6
Thanks!
